rclcpp/include/rclcpp/experimental/subscription_intra_process_buffer.hpp

```diff
@@ -15,50 +15,61 @@
 #ifndef RCLCPP__EXPERIMENTAL__SUBSCRIPTION_INTRA_PROCESS_BUFFER_HPP_
 #define RCLCPP__EXPERIMENTAL__SUBSCRIPTION_INTRA_PROCESS_BUFFER_HPP_
 
-#include <rmw/rmw.h>
-
-#include <functional>
-#include <map>
 #include <memory>
-#include <stdexcept>
 #include <string>
+#include <stdexcept>
 #include <utility>
 
 #include "rcl/error_handling.h"
+#include "rcl/guard_condition.h"
+#include "rcl/wait.h"
 
-#include "rclcpp/any_subscription_callback.hpp"
 #include "rclcpp/experimental/buffers/intra_process_buffer.hpp"
 #include "rclcpp/experimental/create_intra_process_buffer.hpp"
 #include "rclcpp/experimental/subscription_intra_process_base.hpp"
+#include "rclcpp/experimental/ros_message_intra_process_buffer.hpp"
 #include "rclcpp/qos.hpp"
 #include "rclcpp/type_support_decl.hpp"
-#include "rclcpp/waitable.hpp"
-#include "tracetools/tracetools.h"
 
 namespace rclcpp
 {
 namespace experimental
 {
 
 template<
-  typename MessageT,
-  typename Alloc = std::allocator<void>,
-  typename Deleter = std::default_delete<MessageT>
+  typename SubscribedType,
+  typename Alloc = std::allocator<SubscribedType>,
+  typename Deleter = std::default_delete<SubscribedType>,
+  /// MessageT::ros_message_type if MessageT is a TypeAdapter,
+  /// otherwise just MessageT.
+  typename ROSMessageType = SubscribedType
 >
-class SubscriptionIntraProcessBuffer : public SubscriptionIntraProcessBase
+class SubscriptionIntraProcessBuffer : public ROSMessageIntraProcessBuffer<ROSMessageType,
```

Comment:
I'm not sure I understand the logic with the hierarchy in the Subscription IPC class.
We have that:

`SubscriptionIntraProcess` is a
`SubscriptionIntraProcessBuffer` is a
`ROSMessageIntraProcessBuffer` is a
`SubscriptionIntraProcessBase` is a
`Waitable`

Do we really need all these layers?

Reply:
> Do we really need all these layers?

I agree that this is confusing, but I think we do need all of these layers.

So at the base, the whole thing needs to be a `Waitable` because it needs to be waited on by the executors.  We need the `SubscriptionIntraProcessBase` as a non-templated type so that we can store essentially untyped subscriptions in the `IntraProcessManager` (https://github.com/ros2/rclcpp/blob/c8c4851b53de49cc1fe2f3fc3d36ef3e2df8ffee/rclcpp/include/rclcpp/experimental/intra_process_manager.hpp#L319-L320).  The split between `ROSMessageIntraProcessBuffer` and `SubscriptionIntraProcessBuffer` is how the subscriber ultimately communicates whether it is a ROS type or not to the `IntraProcessManager`.  So we need both of those layers to distinguish between those two cases.  And finally, we need `SubscriptionIntraProcess` as the thing that the `Subscription` class interacts with.

I'll note that this hierarchy was largely here already; the only thing this PR does is to add in the `ROSMessageIntraProcessBuffer`.

One thing we *might* be able to do is collapse the `SubscriptionIntraProcess` and `SubscriptionIntraProcessBuffer` classes into one.  Neither of them does too much on its own, and it isn't entirely clear why they are separated.  If you'd like me to try to do that, I'll give that a whirl.  I think we need the rest of the layers.

Reply:
> I'll note that this hierarchy was largely here already; the only thing this PR does is to add in the `ROSMessageIntraProcessBuffer`.

Do you think that it makes sense to rename `ROSMessageIntraProcessBuffer` to something like `SubscriptionROSMessageIntraProcessBuffer`, or something similar? It is just a little odd to me that it's in the hierarchy but doesn't start with `Subscription`, like the other classes after waitable.

>  If you'd like me to try to do that, I'll give that a whirl. I think we need the rest of the layers.

Maybe make an issue and possibly do it in another PR? This isn't that high priority for me.

Reply:
> Do you think that it makes sense to rename ROSMessageIntraProcessBuffer to something like SubscriptionROSMessageIntraProcessBuffer, or something similar? It is just a little odd to me that it's in the hierarchy but doesn't start with Subscription, like the other classes after waitable.

That makes sense to me.  I'll go ahead and rename it (probably to `SubscriptionROSMsgIntraProcessBuffer`, just to shorten it a bit).